### ISSUE TITLE
Cleanup jobs not properly created for TaskChain + KeepOutput=False

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#8723]
More info in the egroup discussion, subject "CERN unmerged area investigation"

Pasting here the important information:
...
From Sid
I've put two lists in [1]. eos_dump.txt contains a directory breakdown of the eos unmerged space, as well as the total (1.45 PB). other_sites.txt contains the size of unmerged at other sites, according to the dynamo consistency tool. There are some caveats as to what goes into the latter (for example, RAW is always ignored), but it should be a good estimate.
[1] http://t3serv001.mit.edu/~snarayan/browse.php?dir=transfers/unmerged_dump
...

and my investigation
...
thanks for the eos dump!
I sorted those directories by data volume and there is around 40 directories with >= 10TB
of unmerged data. Then I picked the worst offender (41 TB):
/store/unmerged/RunIIFall17wmLHEGS/W1JetsToLNu_LHEWpT_150-250_TuneCP5_13TeV-amcnloFXFX-pythia8

and started debugging it from there. I made several notes and checks here and there and you
can see all of that in this text file:
https://www.dropbox.com/s/qvavbp9qv41jprn/debug_store_unmerged_CERN_EOS.txt?dl=0

The workflow creating those output datasets was acquired by 4 different agents (plus a bunch
of ACDC workflows), but I did all this investigation in one agent only - which is fine.

In short, I think it's safe to acknowledge there is a bug in WMAgent regarding Cleanup job
creation for TaskChain workflows with KeepOutput=False (staging the unmerged output but
not merging neither injecting into DBS/PhEDEx). This seems to affect only the specific output
that is used as input for further processing tasks though (see LHE check on the text doc).

Note, however that the workflow I used for this investigation was announced yesterday, so it
shall be archived today and all the agent information gone (that also means that unmerged
path will get unlocked by Unified and cleaned up by Dynamo soon, I think).

I'm creating a WMCore issue with the content of this email, but that shouldn't stop us from
safely releasing /store/unmerged space in a different way. I bet this cleanup job creation
has a nasty logic and it's not going to be straight forward to get it fixed.

PS.: Another thing that I cannot explain is why around 15k unmerged files were not processed
by a further Processing job (186478 out of 203950 files)!?!
...

**Comment by amaltaro** [comment]
@ticoann before we start working to fix it, can you please go through my text file and double check my steps, especially the sql queries just to make sure I didn't make any mistake and we really have this issue in the agent code.

**Comment by amaltaro** [comment]
Just to make it harder to debug... it doesn't seem to happen to all such requests. Looking at this workflow (already in completed)
prozober_task_HIG-RunIIFall17wmLHEGS-00746__v1_T_180714_121808_9488

with Task1 and Task2 KeepOutput=False, I cannot find any unmerged leftovers in EOS, even though looking at vocms0257 database it tells me not all unmerged files have been assigned to cleanup jobs.

There are 74 unmerged GEN-SIM-RAW files created:
```
select count(distinct lfn) from wmbs_file_details where lfn like '/store/unmerged/RunIIFall17DRPremix/GluGluToHiggs0PHf05ph0continToZZTo2e2mu_M125_10GaSM_13TeV_MCFM701_pythia8/GEN-SIM-RAW/PU2017_94X_mc2017_realistic_v11-v2/%';
COUNT(DISTINCTLFN)
------------------
		74
```

while only 70 of them have been assigned to a Cleanup job (74 assigned to Processing jobs)
```
SELECT wmbs_sub_types.name, count(DISTINCT wmbs_job.id), count(DISTINCT wmbs_file_details.id)
  FROM wmbs_file_details
    INNER JOIN wmbs_job_assoc ON wmbs_job_assoc.fileid = wmbs_file_details.id
    INNER JOIN wmbs_job ON wmbs_job.id = wmbs_job_assoc.job
    INNER JOIN wmbs_jobgroup ON wmbs_jobgroup.id = wmbs_job.jobgroup
    INNER JOIN wmbs_subscription ON wmbs_subscription.id = wmbs_jobgroup.subscription
    INNER JOIN wmbs_sub_types ON wmbs_sub_types.id = wmbs_subscription.subtype
    INNER JOIN wmbs_workflow ON wmbs_workflow.id = wmbs_subscription.workflow
    INNER JOIN wmbs_location_pnns ON wmbs_location_pnns.location = wmbs_job.location
    INNER JOIN wmbs_pnns ON wmbs_pnns.id = wmbs_location_pnns.pnn
  WHERE 
    wmbs_file_details.lfn like '/store/unmerged/RunIIFall17DRPremix/GluGluToHiggs0PHf05ph0continToZZTo2e2mu_M125_10GaSM_13TeV_MCFM701_pythia8/GEN-SIM-RAW/PU2017_94X_mc2017_realistic_v11-v2/%'
    GROUP BY wmbs_sub_types.name;
NAME			COUNT(DISTINCTWMBS_JOB.ID) COUNT(DISTINCTWMBS_FILE_DETAILS.ID)
--------------- -------------------------- -----------------------------------
Cleanup 		31						    70
Processing		55						    74
```

**Comment by amaltaro** [comment]
BTW, we have two templates exercising that same TaskChain setup:
WMCore/test/data/ReqMgr/requests/Integration/TaskChain_MC.json
WMCore/test/data/ReqMgr/requests/Integration/TaskChain_DropOutputs.json

and none of them have uncleaned unmerged files.

**Comment by vlimant** [comment]
@amaltaro @ticoann can you please update this and come up with something quite soon about this

**Comment by ticoann** [comment]
Hi Jean-Roch, we are still investigating. Alan found out that files not which is not cleaned are actually used for the failed jobs. (checked prozober_task_HIG-RunIIFall17wmLHEGS-00746__v1_T_180714_121808_9488). Which is expected since we can't clean up unmerged input files which is failed unless ACDC is created and succeeded which was original design.  We are trying to check some other workflows whether that is consistent and actually that is the sole cause of unmerged files not deleted. If it is the case, I don't think that is the bug. We have have some other way to clean up those files fast. (Dynamo delete those files eventually with the lock condition, correct?)

**Comment by bbockelm** [comment]
Another suggestion -

Can we tally up the breakdown of unmerged size at CERN by workflow priority?  That would allow a different kind of sanity check.

If we see that 80% of unmerged space goes to low-priority workflows -- well, there's not much we can do other than change how we interpret priority.

**Comment by vlimant** [comment]
@ticoann @amaltaro is there a way to list all unmerged files that are still in use from the agent, so that we dump this somewhere for dynamo to use and clean files that were not cleaned by the agent in the first place, but that would live in a directory that is still protected

**Comment by ticoann** [comment]
@bbockelm, Here is the list of unmerged area sorted by Alan's document above. 
request and priority related to that.

The structure is
```
umerged area  size (TB)
     request name,  request status,  initial priority, request priority 
```

https://www.dropbox.com/s/2k1hnq505zzp35y/umerged.txt?dl=0

The script is just checking whether certain workflow is contributing data in given unmerged space.
So it doesn't mean it is responsible for all the space. It might not contribute at all. It also only includes the current active requests (not archived) so it might be possible actual contribution of data is from other workflows which already archived (and space might be cleaned up).

So it is just a rough view what kind of priority might contribute on above unmerged space.

That said, I am not sure what is considered low priorities. Comparing current active workflows, I think most of them are below the median  among currently running workflows.

**Comment by ticoann** [comment]
> is there a way to list all unmerged files that are still in use from the agent

I think we can get the unmerged acdc inputs from db. But I am not sure I follow the logic about cleaning. You want to move those files on specific location so dynamo can deleted later? 
Sorry if I misunderstood.

**Comment by bbockelm** [comment]
In terms of priorities - let's call anything less than 100k at CERN "low priority".

Can we make this simpler by only looking at files WMAgent knows about?  We can separate out files that exist but not known to the agent (although it'd be an interesting sub-thread to attempt to count these!).

**Comment by ticoann** [comment]
One other thing is we still not sure whether Agent has a bug or not. I will try to test at least few more workflows which doesn't create the clean up jobs falls into failed job case.

**Comment by amaltaro** [comment]
Let me jump into this discussion. I think we're trying to solve it the hard way, to be frank.
Fetching unmerged files known by the agent that isn't anymore needed by any job is a total overkill and waste of time in my opinion. Keep in mind that we don't run a single agent and there is not a single database for fetching this info (besides the ACDCServer, but that would be terribly expensive).

My suggestion still stands, why don't we reduce the waiting period of *unlocked* LFNBases in dynamo and let it clean files older than X days (why not 1 day?) instead of the current 28 days. I think Dan mentioned there would be a few hundred TBs that would be cleaned right away. It doesn't need to be anything permanent nor anything for all sites, just to CERN such that we can start using those 70k resources sooner than later.

PS.: Honestly, I also fail to see what role the workflow priority is playing here. Are you considering aborting/cleaning up and running that (low) priority workflow elsewhere?

**Comment by bbockelm** [comment]
That's why I specifically only was interested in looking at one agent - I didn't want to try to construct a global picture.

The role the workflow priority plays here is to help understand why there may be so many files sitting around.  Lots of unmerged files from low-priority workflows are "understood" more than unmerged files from high-priority files.  Are the majority of the files "understood" or "not understood"?

Basically, I'm looking for two numbers -- for agent XYZ, the # of TBs of unmerged data associated with CERN above prio 100k and the number below 100k.

I'm not thinking about any concrete action with these yet, just trying to quantify the different causes.

I agree that understanding unlocked LFNs are important.  We should probably file a separate ticket for that because they are important but a different topic.

**Comment by thongonary** [comment]
I just looked at the list that @ticoann provided, and it struck me that the first workflow that is taking 41 TB is not even running 
https://cmsweb.cern.ch/reqmgr2/fetch?rid=pdmvserv_task_HIG-RunIIFall17wmLHEGS-01778__v1_T_180521_202854_6570
Is this possible? I thought the unmerged area is only used for the intermediate unmerged output datasets?

**Comment by ticoann** [comment]
> Is this possible? I thought the unmerged area is only used for the intermediate unmerged output datasets?

Hi Thong, That is what I was explaining above.

> The script is just checking whether certain workflow is contributing data in given unmerged space.
> So it doesn't mean it is responsible for all the space. It might not contribute at all. It also only includes the current active requests (not archived) so it might be possible actual contribution of data is from other workflows which already archived (and space might be cleaned up).

For the particular space, workflow responsible is already archived. pdmvserv_task_HIG-RunIIFall17wmLHEGS-01115__v1_T_180415_203803_4286

**Comment by thongonary** [comment]
OK I take that single workflow does not take the whole 41 TB, but how come it contributes _any_ data to the unmerged space given that it's not running? 

I also agree with Alan's suggestion of reducing the waiting period for unlocked to 1 day -- from the list I see that most of the top spaces are having only 0 to 2 active workflows, so the majority of the unmerged space comes from the unlocked data of archived workflows that are waiting to be deleted.

**Comment by vlimant** [comment]
https://dmytro.web.cern.ch/dmytro/cmsprodmon/workflows.php?prep_id=task_HIG-RunIIFall17wmLHEGS-01115 has created files that the presence of https://cmsweb.cern.ch/reqmgr2/fetch?rid=pdmvserv_task_HIG-RunIIFall17wmLHEGS-01778__v1_T_180521_202854_6570 locks for dynamo

**Comment by amaltaro** [comment]
Can I suggest we wait one day more to complete this investigation. Seangchan and I are almost sure we found the reason for these leftovers, we "just" need to do some extra checks before coming up with that solution. Part of the conclusion is on Seangchan's comment:
https://github.com/dmwm/WMCore/issues/8723#issuecomment-411125499

and if that's really true, it's not really a bug but a feature of ACDCs :-) Workflow priority has no effect as well, it's all a matter of "a failed processing job that was reading an unmerged file", thus this unmerged file is put in the acdcserver and we take that file out of any cleanup jobs (such that an ACDC can run on it).

**Comment by vlimant** [comment]
hey @amaltaro , if (and only if) there is a bug somewhere that left >1.4PB of unmerged file at CERN (all other places at T1/T2 will run into critical situation at some point) the heavy query on wmbs/agent to get an actual list of protected files, and take evasive action on eos (remove all old files not explicitly protected by that list) is something we have to do to mitigate the issue.

as for lowering down to 1 day, it might be too aggressive (if the list does not update often enough), but 28 is way too much ; has anyone contacted directly transfer team to make this change?

**Comment by bbockelm** [comment]
Hi all,

I'm still really uncomfortable sweeping a petabyte under the rug, shrugging our shoulders, and saying "WMAgent leaks ten terabytes a day, but Dynamo will clean it up".

We have identified one or two possible causes.  Can we estimate what percentage of the files on disk these account for?

If we are chasing files that can be processed by a future ACDC - can we create a monitoring view of this metric?

Brian

**Comment by amaltaro** [comment]
After changing this message a few times before posting it, these were my findings while analyzing another sample workflow:
pdmvserv_task_HIG-RunIIFall17wmLHEGS-01116__v1_T_180418_043017_9486

somehow what's reported by WMStats does not match what is in the ACDC Server. I believe because we had a redeployment campaign (1.1.12) between April and May and that data got lost.
I guess that's expected, but if Seangchan can confirm(?) If it's not expected, we might follow that up in another issue not to spam too much here.

Ok, then adapting and running an old script from 2015
https://github.com/amaltaro/scripts/blob/master/checkACDCCollections.py

and querying the ACDC server for this workflow above, I confirm that CERN EOS has almost the same number of unmerged files in its storage. Check the output of that script against a few eoscms queries:
http://amaltaro.web.cern.ch/amaltaro/forWMCore/Issue_8723/pdmvserv_task_HIG-RunIIFall17wmLHEGS-01116__v1_T_180418_043017_9486

so the only file leak (well, at least that I cannot explain now) happened on the LHE datatier, around 100 files left in the storage, instead of only 3.

> I'm still really uncomfortable sweeping a petabyte under the rug, shrugging our shoulders, and saying "WMAgent leaks ten terabytes a day, but Dynamo will clean it up".

Thus, I'd look at it from another angle. When ACDC was designed it was counting on the site cleanup policy to clean any files meant to be processed by ACDC workflows, which were not really ACDC'ed. 
It was a flaw in the design and now we need to come up with a solution for such cases, which I still don't have and likely won't be simple.

> as for lowering down to 1 day, it might be too aggressive (if the list does not update often enough), but 28 is way too much ; has anyone contacted directly transfer team to make this change?

Call it a week then. Dan said the cleanup script runs only once a week anyways, so if it's a day it will be cleaning data older than 1 to 8 days, if a week data older than 8 to 15 days.
What is the actual concern with this file age here? Are there any cases where a workflow is announced and requestors come back asking for the final lumis to be recovered (if so, how often?)?
Otherwise, how the list is built looks pretty safe to me:
https://github.com/dmwm/WMCore/blob/6c0c1283374b25a9d0d9a3ffbc59af3777c29a13/src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py#L121

**Comment by ticoann** [comment]
Thanks Alan for the investigation.

> somehow what's reported by WMStats does not match what is in the ACDC Server. I believe because we had a redeployment campaign (1.1.12) between April and May and that data got lost.

What is not matching? If the number of failed jobs, yes, it is possible between the redeployment. But it shouldn't be many job report loss unless there was replication problem.  We are redeploying agents after all the jobs are finished and upload to DBS and PhEDEx. Although we don't check replication status before redeploy the agent, it all the reports should be reported by then.  Other possibility is we just lost the report update in local couch db. When failed job report comes back we try to update once and if it fails we don't retry. (That was by design with the assumption that local couch is reliable and we can afford to lose some reports for monitoring).

> What is the actual concern with this file age here? Are there any cases where a workflow is announced and requestors come back asking for the final lumis to be recovered (if so, how often?)?

Just a little more detail about what Alan said. As far as I understood when workflow is announced by unified, it change the status of all the related workflow (ACDC, cloned, recovery) to announced.
Then the umerged lfn base is remove the protected list. I haven't even thought about the use cases what Alan mentioned above. It the concern is whether protected lfn list is constructed correctly as intended we should double check and make sure.

**Comment by vlimant** [comment]
@amaltaro "What is the actual concern with this file age here?" just making sure the file is not newer than when the list of protected directory was updated last : a week would do.

**Comment by vlimant** [comment]
@amaltaro  regarding https://github.com/dmwm/WMCore/blob/6c0c1283374b25a9d0d9a3ffbc59af3777c29a13/src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py#L121 could have assigned and acquired removed from the list of statuses to lock on, since those guys have not yet created anything by any mean (https://github.com/CMSCompOps/WmAgentScripts/commit/8f8772362a8b4e062f9b65127ed0320a8d2a11b7)

**Comment by vlimant** [comment]
concerning the issue, there has been some back and forth and I want a clean and definite answer @amaltaro @ticoann on whether files are left behind from task of taskchain with KeepOutput=False.
Is this a "yes files are left behind" (bug in wmagent), or "no, files are properly cleaned up, if not the source of a failed job" (no bug, and those files are required). You can take a couple of days as per https://github.com/dmwm/WMCore/issues/8723#issuecomment-411227958 request

**Comment by ticoann** [comment]
> could have assigned and acquired removed from the list of statuses to lock on, since those guys have not yet created anything by any mean (CMSCompOps/WmAgentScripts@8f87723)

There is always race condition for when status is queried and and actual deletion happens (status change  in mean time). However, I think it is safe if we wait at least 1 day before deletion.
I was wondering whether it cause more problem including those? If that is the problem having more list,  we can remove those status. ('assigned', 'acquired')

**Comment by bbockelm** [comment]
@amaltaro - this looks like great progress!

To make sure I understand:
1.  These are files known to the ACDC server
2.  We've decided (for whatever reason, doesn't matter) to not run an ACDC for this workflow.
3.  ACDC is never informed of this fact.  Even if it was, there's no ACDC "cleanup" mechanism that it can kick off.

Does the system know when we won't actually run the ACDC?  I always feel better stating the files it is safe to delete as opposed to locking files.

**Comment by ticoann** [comment]
> Does the system know when we won't actually run the ACDC? I always feel better stating the files it is safe to delete as opposed to locking files.

When original workflow is announced which means data ops are ok with  data produced already and unified announce all the related ACDCs (@vlimant Jean-Roch can confirm). Than we are safe to delete the files except for some reason people decided to collect more data from already announce workflow. (As Alan was mentioning). However ACDC data will be cleaned up (not the actual files but metadata from couch db), once workflow is archived (There is a grace period) so it is hard to even produce new data.

We provide REST api for protected lfn base: (means the unmerged files which might be still in use)
https://cmsweb.cern.ch/wmstatsserver/data/protectedlfns

Once a workflow is announced the unmerged lfn base (for output data) from the workflow will be removed from above API result.

**Comment by amaltaro** [comment]
Trying my best to give a clean reply, since we've a few sort of diff topics on going now :)

About producing the protectedlfns list, I think it doesn't hurt us much if we keep those two status (assigned, acquired). That's also a safeguard in case the request transition thread goes down or doesn't run for some unexpected time.

About the cleanup policy, I suggest we follow that up in another thread. One week should do the job for now, probably not in a year or two though. We might also push for the new policy not only for CERN, but everywhere. Anyhow..

@bbockelm yes, you summarize well the issue. As Seangchan mentioned, IF ACDC is not created, then we rely on the sites (now DM) to clean those files. Once the workflow gets archived (or announced) in reqmgr2 - I just checked the code, apparently there is no grace period - all the ACDC information is cleaned up.

Jean-Roch, I'm going to perform the same check on a couple of other requests, but I believe (hopefully) the conclusion won't change.

**Comment by amaltaro** [comment]
Ok, I've done ACDC x EOS comparisons for 4 more requests in completed (but with ACDCs already running):
pdmvserv_task_SMP-RunIISummer15wmLHEGS-00227__v1_T_180416_194004_2476
prozober_task_SMP-RunIISummer15wmLHEGS-00226__v1_T_180611_192517_1020
pdmvserv_task_HIG-RunIIFall17wmLHEGS-01122__v1_T_180415_203643_8440
pdmvserv_task_HIG-RunIIFall17wmLHEGS-01114__v1_T_180415_203757_5745

and in general the number of unmerged files in the ACDCServer is very close to the amount of files under CERN EOS paths. If you want, you can see my notes and outputs in the dropbox file I shared in my first comment.

I used this script for the EOS checks:
https://github.com/amaltaro/scripts/blob/master/checkEOSPath.py

In my opinion, WMAgent is not leaking files and those leftovers belong to jobs that failed, so the agent doesn't delete because we might run ACDC on them. As already explained above.

PS.: the only case I see more files than expected is for the LHE tier. However, it could be that there were cleanup jobs failing (wmstats got rotated, so we can't be sure how many failed for these old workflows). To be investigated again in the near future...

**Comment by vlimant** [comment]
thanks

**Comment by ticoann** [comment]
closing the issue since it seems agent doing what it is supposed to do.
